sympy/core/power.py

```diff
@@ -1662,7 +1662,11 @@ def mul(d1, d2):
             g = (b - f).simplify()/f
             _, d = g.leadterm(x)
             if not d.is_positive:
-                raise NotImplementedError()
+                if limit(g/x**maxpow, x, 0) == 0:
```

Comment:
Is there a test for this case?

Reply:
It is needed in - 
```py
def test_issue_13332():                                                         
    assert limit(sqrt(30)*5**(-5*x - 1)*(46656*x)**x*(5*x + 2)**(5*x + 5*S.Half) *
                (6*x + 2)**(-6*x - 5*S.Half), x, oo) == Rational(25, 36)
```
Here, at an intermediate stage, we get (on printing g, g.is_zero and self (on next line))
```py
-(_p + log(531441*(2*_w + 5)**5/(3125*(_w + 3)**6)))/(-log(_w) - 5*log(5) + 6*log(3)) None
1/(_p/_w - 6*log(1 + 3/_w)/_w + 5*log(2 + 5/_w)/_w - 6*log(2)/_w + 6*log(6)/_w)
```
Since g.is_zero is returning None, we need to handle this case separately.

Maybe a better fix would be changing - https://github.com/sympy/sympy/blob/0f782091845fbb52b26c3d55fb50240c260054b4/sympy/core/power.py#L1628 to `if g.is_zero is not False`

Reply:
> if g.is_zero is not False

Have you run tests with this?

Reply:
I tried running tests, and it fails for examples like `g = (-_x + tan(_x))/_x`

Reply:
> -(_p + log(531441*(2*_w + 5)**5/(3125*(_w + 3)**6)))/(-log(_w) - 5*log(5) + 6*log(3))

It seems that `_p` should have been replaced by `-log(_w)` (because `x, logx` are `_w, -_p`). Hence these substitutions act in the wrong direction: https://github.com/sympy/sympy/blob/ea3aedd21c2d6b0a0e1021382ad1c45441ab0060/sympy/core/power.py#L1599-L1601

This should probably be something like

        if logx is not None:
            b.subs(logx, log(x))

Reply:
Changing it to - 
```py
        if logx is not None:                                                    
            b = b.subs(logx, log(x))                                            
            self = b**e
```
causes a lot of test failures. `_p` is indeed replaced by `-log(_w)` in the expressions entering this try block. The example I mentioned now raises ValueError in `Mul._eval_nseries`, and further, some mrv computations in the tests also raise the same error from here.

For `mrv(exp(-x + exp(-x)*exp(-x*log(x))), x)`, g and self are `-1/log(_w)` and `1/(-log(1/_w)/_w - 1/_w)`, and it raises `NotImplementedError` at the original try block in this review (L1665 in power.py)

Some places in `test_aseries` also enter an infinite recursion. I tried debugging and looking around but have no clue why so many tests are failing in different fashions. Perhaps the code requires `_p` to remain in the expression.

Reply:
What will happen if logx is not replaced either way?

Reply:
I tried going through several examples and it looks like the original expression (self) never has -_p (logx) term. It only has terms in _w (and log(_w)). So, substituting logx for log(x) essentially does nothing. Maybe it was intended to substitute log(x) by logx.

I tried changing it to `b = b.subs(log(x), logx)`, and in this case, the limit example works fine but there is one mrv computation failure in the tests (mentioned in my previous comment). This is because b.subs(log(x), logx) does not substitute `log(1/_w)` for `_p`, but the current implementation does.

Reply:
There doesn't seem to exist any proper documentation on how `logx` should be treated by the core methods. In the expansion

    f = c0*w^e0 + c1*w^e1 + ... + O(w^en),       where e0<e1<...<en, c0!=0

given in the docstring of `gruntz`, the coefficients `ci` are only "constant-like", i.e., they may depend on slowly varying functions of `w` like `log(w)`. I think that is why `gruntz` makes `log(w)` a real constant before calling the series expansion code. But I'm not clear when the code methods should make substitutions. If that was not expected, then obviously there would be no need to give `logx` as a call parameter.

Reply:
It seems that the `logx` parameter should be added to all leading term methods. Otherwise the result will not be the same as the first term of a series expansion:
```
In [9]: w, p = symbols('w p', positive=True)                                    

In [10]: f = log(1 + 3/w)/w                                                     

In [11]: f.as_leading_term(w)                                                   
Out[11]: 
   ⎛3⎞
log⎜─⎟
   ⎝w⎠
──────
  w   

In [12]: f.series(w, 0, 2, logx=-p)                                             
Out[12]: 
p + log(3)   1   w     ⎛ 2⎞
────────── + ─ - ── + O⎝w ⎠
    w        3   18        
```

Reply:
It is possible to emulate the substitution by changing `log(1 + 3/w)` to `p + log(w + 3)`, etc. Then the leading term `f` of `b` becomes
```
In [13]: b = p/w - 6*(p + log(w + 3))/w + 5*(p + log(2*w + 5))/w - 6*log(2)/w + 
    ...: 6*log(6)/w                                                             

In [14]: b.as_leading_term(w)                                                   
Out[14]: 
-6⋅log(3) - 6⋅log(2) + 5⋅log(5) + 6⋅log(6)
──────────────────────────────────────────
                    w                     
```
which is different from the value computed above.

Reply:
So, all _eval_as_leading_term methods should be changed to something like - 
```py
def _eval_as_leading_term(self, x, logx=None, cdir=0):
    if logx is not None:
       c, ex = symbols('c, ex', cls=Wild, exclude=[x])
       b = b.replace(log(c*x**ex), log(c) + ex*logx)
       self = b**e
    # Continue original code
```
and calls to `_eval_as_leading_term`, `as_leading_term` and `leadterm` methods should also take a logx parameter, when they care being called inside `_eval_nseries` or `_eval_as_leading_term`.

Is this the correct way to go about solving this?

Reply:
It should suffice to add the `logx` parameter to each call of `leadterm` or `as_leading_term`, and then handle the substitution only in `log._eval_as_leading_term`.

Reply:
I am trying to handle the substition in `Expr.as_leading_term` as - 
```py
       ...
        if x not in self.free_symbols:
            return self

        expr = self
        if isinstance(expr, log) and logx is not None:
            arg = expr.args[0].together().factor()
            expr = expand_log(self.func(arg))
            expr = expr.subs(log(x), logx)
        obj = expr._eval_as_leading_term(x, logx=logx, cdir=cdir)

        if obj is not None:
        ...
```
But it keeps throwing recursion errors for many limits. It is a bit confusing and I am not able to see the source of error, so it would be very helpful if you could point out the error. I have also changed all _eval_as_leading_term methods to take logx as an argument (default None), and the examples you gave above are working with this - 
```py
In [1]: w, p = symbols('w p', positive=True)

In [2]: f = log(1 + 3/w)

In [3]: f.as_leading_term(w, logx=-p)
Out[3]: p + log(3)

In [4]: f.series(w, 0, 2, logx=-p)
Out[4]: 
         w        ⎛ 2⎞
log(3) + ─ + p + O⎝w ⎠
         3            

```

Reply:
I could not find the original example in #13332. Is the issue number correct? Maybe this test (and the fix) should be omitted for now.

Reply:
The example is correct I believe - 
```bash
$ cat a.py 
import sympy as sp

def stirling(n):
    return sp.sqrt(2 * sp.pi * n) * (n / sp.E)**n

def binom_stirling(n, k):
    return stirling(n) / (stirling(k) * stirling(n - k))

n, k = sp.symbols('n k')

r0 = sp.simplify(binom_stirling(2*(3*k), k))
r1 = sp.simplify(binom_stirling(2*(3*k+1), k))
r = sp.simplify(r0 / r1)

print(r, k, sp.oo)

$ python a.py 
sqrt(30)*5**(-5*k - 1)*(46656*k)**k*(5*k + 2)**(5*k + 5/2)*(6*k + 2)**(-6*k - 5/2) k oo
```

Reply:
Okay, after a lot of debugging, I found another working diff, without this limit check - 
```diff
diff --git a/sympy/core/add.py b/sympy/core/add.py
index 3294b4ba79..09df1b5463 100644
--- a/sympy/core/add.py
+++ b/sympy/core/add.py
@@ -1005,7 +1005,7 @@ def _eval_as_leading_term(self, x, cdir=0):
         except TypeError:
             return expr
 
-        new_expr = new_expr.together().factor()
+        new_expr = new_expr.together().trigsimp().factor()
         if new_expr.is_Add:
             new_expr = new_expr.simplify()
 
diff --git a/sympy/core/power.py b/sympy/core/power.py
index 0e58ceb01a..08567ce78a 100644
--- a/sympy/core/power.py
+++ b/sympy/core/power.py
@@ -1659,14 +1659,10 @@ def mul(d1, d2):
             else:
                 raise NotImplementedError()
         if not d.is_positive:
-            g = (b - f).simplify()/f
+            g = g.simplify()
             _, d = g.leadterm(x)
             if not d.is_positive:
-                if limit(g/x**maxpow, x, 0) == 0:
-                    # g has higher order zero
-                    return f**e + e*f**e*g  # first term of binomial series
-                else:
-                    raise NotImplementedError()
+                raise NotImplementedError()
 
         gpoly = g._eval_nseries(x, n=ceiling(maxpow), logx=logx, cdir=cdir).removeO()
         gterms = {}
```

However, I am not sure about the fact of adding simplification functions to new_expr in Add as and when we are finding failing test cases. In particular, trigsimp is needed for `new_expr = -2*(sin(a)**2 + cos(a)**2 - 1)`, which does not return 0 on `together()` or `factor()` calls.

Reply:
At this point, it might make sense to replace all these various calls (after which again there is a chance of simplify being called), with just one call to `simplify()` directly. However, that also comes with it's own challenges and issues, mainly - 
```py
In [4]: limit(((1 + ((S(2)/3)**(x + 1)))**(2**x))/(2**((S(4)/3)**(x - 1))), x, oo) #Evaluates to oo in master
Out[4]: 
    ⎛                         ⎛ x⎞⎞
    ⎜     x - 1               ⎝2 ⎠⎟
    ⎜ -4/3      ⎛   x + 1    ⎞    ⎟
lim ⎝2         ⋅⎝2/3      + 1⎠    ⎠
x─→∞                               

In [5]: a, b, c, x = symbols('a b c x', positive=True)

In [6]: limit((a + 1)*x - sqrt((a + 1)**2*x**2 + b*x + c), x, oo)
Out[6]: 
      ⎛       ______________    ⎞
      ⎜      ╱  2               ⎟
∞⋅sign⎝a - ╲╱  a  + 2⋅a + 1  + 1⎠

```
Maybe the current block of `new_expr` is actually masking underlying issues with other leading term methods. What are your thoughts on this, and how should I proceed? Would it suffice to apply the diff I mentioned?

Reply:
> new_expr = new_expr.together().trigsimp().factor()

This is beginning to look complicated. It seems that `.is_zero` would return None for potentially zero expressions. I would try something like this instead of many explicit simplifications as above:

        iszero = new_expr.is_zero
        if iszero == None:
            new_expr = new_expr.simplify()
            iszero = new_expr.is_zero
        if iszero == True:
            # simple leading term analysis gave ...

Reply:
Since the `_mexpand` leads to expansion of things like `(a + 1)**2`, a `factor` is still needed, for `limit((a + 1)*x - sqrt((a + 1)**2*x**2 + b*x + c), x, oo)` (my previous comment)